fuseL+fuseR

safareli opened the pull request:
I have expanded pattern matches of fuse and used it for implementation.

safareli committed:
fix warning

safareli committed:
use safer implementations for to* functins

without this change toProcess and toConsumer are *incorrect* as this will compile:
``` purs
x :: forall m. Monad m => Transducer Void Void m Unit
x = do
  res <- awaitT
  case res of
    Just void' -> absurd void'
    Nothing -> pure unit
```

safareli committed:
fuseL+fuseR

I have expanded pattern matches of fuse and used it for implementation.

safareli committed:
add these version of fuse

blinky3713 commented:
Cool, let's go ahead and get some tests in here and fix #4 while we're at it. Since we're adding coroutines with specific termination criteria, let's make sure they all work, and some in combination .

it's interesting, because these have different meanings

```
(t1 `fuseL` t2) `fuseR` t3 /= t1 `fuseL` (t2 `fuseR` t3)
```

safareli committed:
Merge branch 'fusel-r' into safe

safareli committed:
use safer implementations for to* functins

without this change toProcess and toConsumer are *incorrect* as this will compile:
``` purs
x :: forall m. Monad m => Transducer Void Void m Unit
x = do
  res <- awaitT
  case res of
    Just void' -> absurd void'
    Nothing -> pure unit
```

safareli committed:
add some coroutine style functins:
runProcess, producer, consumer add transformForever

use untillJust whileJust from ps-tailrec

test actually performs assertion and terminates
